cmd/frontend/internal/pkg/langservers/config.go

```
@@ -53,23 +54,41 @@ func State(language string) (ConfigState, error) {
 	return StateNone, nil
 }
 
-func setDisabledInGlobalSettings(language string, disabled bool) error {
+func setDisabledInGlobalSettings(ctx context.Context, language string, disabled bool) error {
 	settings, err := db.Settings.GetLatest(context.Background(), api.ConfigurationSubject{Site: true})
 	if err != nil {
 		return err
 	}
+	var contents string
+	var id *int32
+	var authorUserID int32
 	if settings == nil {
-		return nil
+		contents = "{}"
+		// HACK: Global settings are nil (probably because this is a brand new
+		// instance), so there's no existing author user ID to reuse. Just take
+		// any user's ID. The author isn't shown anywhere, anyway.
+		users, err := db.Users.List(ctx, &db.UsersListOptions{
+			LimitOffset: &db.LimitOffset{Limit: 1},
+		})
+		if err != nil || len(users) == 0 {
+			log15.Warn("unable to obtain a user ID to edit global settings and enable/disable a language server, so global settings may be out of sync with `langservers` in site config")
+			return nil
```

I believe you should return an error here instead.

Doing it this way via logging means that the global settings and site settings will become out of sync, which is ugly and requires the user to fix it. By returning an error, the two do not get out of sync and instead the language just is not enabled at all. This seems like a better outcome.

---

I originally had it throw an error, but it's probably easier for the user to edit global settings than to edit site config plus manually start/stop docker containers. I'd rather the edits to global settings not prevent the admin from being able to start/stop containers.

---

Here is the difference in behavior:

If we return an error:

- The language server will not be started.
- The extension will not be enabled in the global settings.
- It will appear to the user as though code intelligence was simply not enabled automatically.
- To fix the situation, the user can just visit the code intelligence page and click `Enable`. It'll then just work.

If we log an error:

- The language server will be started.
- The extension will not be enabled in the global settings.
- It will appear to the user that code intelligence does not work when interacting with code, but viewing the code intelligence page will indicate the language server is enabled and running healthy.
- To fix the situation, the user will have to edit global settings OR first Disable in the code intelligence page and then enable again for it to work.

For me, the error states and fix the user would perform seem much cleaner if we return an error.

---

> To fix the situation, the user can just visit the code intelligence page and click Enable. It'll then just work.

But clicking Enable calls this function, and when this function fails, Enable will abort.

---

Hopped on a call, @slimsag convinced me that it's ok to couple the DB and the code intelligence page (sometimes it'll be a transient problem, and if not then the user has more serious problems to deal with, namely that the DB is down).